source/common/config/watch_map.cc

```diff
@@ -0,0 +1,171 @@
+    const absl::flat_hash_set<Watch*>& interested_in_r =
+        watchesInterestedIn(name_getter.resourceName(r));
+    for (const auto& interested_watch : interested_in_r) {
+      per_watch_updates[interested_watch].Add()->CopyFrom(r);
+    }
+  }
+
+  // We just bundled up the updates into nice per-watch packages. Now, deliver them.
+  for (auto& watch : watches_) {
+    auto this_watch_updates = per_watch_updates.find(watch);
+    if (this_watch_updates == per_watch_updates.end()) {
+      // This update included no resources this watch cares about - so we do an empty
+      // onConfigUpdate(), to notify the watch that its resources - if they existed before this -
+      // were dropped.
+      watch->callbacks_.onConfigUpdate({}, version_info);
+    } else {
+      watch->callbacks_.onConfigUpdate(this_watch_updates->second, version_info);
+    }
+  }
+}
+
+void WatchMap::onConfigUpdate(
+    const Protobuf::RepeatedPtrField<envoy::api::v2::Resource>& added_resources,
+    const Protobuf::RepeatedPtrField<std::string>& removed_resources,
+    const std::string& system_version_info) {
+  // Build a pair of maps: from watches, to the set of resources {added,removed} that each watch
+  // cares about. Each entry in the map-pair is then a nice little bundle that can be fed directly
+  // into the individual onConfigUpdate()s.
+  absl::flat_hash_map<Watch*, Protobuf::RepeatedPtrField<envoy::api::v2::Resource>> per_watch_added;
+  for (const auto& r : added_resources) {
+    const absl::flat_hash_set<Watch*>& interested_in_r = watchesInterestedIn(r.name());
+    for (const auto& interested_watch : interested_in_r) {
+      per_watch_added[interested_watch].Add()->CopyFrom(r);
+    }
+  }
+  absl::flat_hash_map<Watch*, Protobuf::RepeatedPtrField<std::string>> per_watch_removed;
+  for (const auto& r : removed_resources) {
+    const absl::flat_hash_set<Watch*>& interested_in_r = watchesInterestedIn(r);
+    for (const auto& interested_watch : interested_in_r) {
+      *per_watch_removed[interested_watch].Add() = r;
+    }
+  }
+
+  // We just bundled up the updates into nice per-watch packages. Now, deliver them.
+  for (const auto& added : per_watch_added) {
+    const Watch* cur_watch = added.first;
+    auto removed = per_watch_removed.find(cur_watch);
+    if (removed == per_watch_removed.end()) {
+      // additions only, no removals
+      cur_watch->callbacks_.onConfigUpdate(added.second, {}, system_version_info);
+    } else {
+      // both additions and removals
+      cur_watch->callbacks_.onConfigUpdate(added.second, removed->second, system_version_info);
+      // Drop the removals now, so the final removals-only pass won't use them.
+      per_watch_removed.erase(removed);
+    }
+  }
+  // Any removals-only updates will not have been picked up in the per_watch_added loop.
+  for (auto& removed : per_watch_removed) {
+    removed.first->callbacks_.onConfigUpdate({}, removed.second, system_version_info);
```

Comment:
So, we might have two `onConfigUpdate`s here? This seems like something that we could collapse into one in all cases.

Reply:
There will not be; see the comment. The .erase() earlier ensures that.